### ISSUE TITLE
BUG: `full_like` and `full` accept array `fill_value`

### ISSUE THREAD

**Issue by lucascolley** [data-apis/array-api-strict#55]
```python
In [1]: import array_api_strict as xp

In [2]: xp.full_like(xp.asarray(0), xp.asarray(1))
Out[2]: Array(1, dtype=array_api_strict.int64)
```

[Spec](https://data-apis.org/array-api/latest/API_specification/generated/array_api.full_like.html):

```
fill_value (Union[bool, int, float, complex]) – fill value.
```

**Comment by asmeurer** [comment]
This seems like it should be fixed in the standard. Getting a fill value from an array seems like it wouldn't be uncommon.

**Comment by kgryte** [comment]
Torch only [accepts a scalar](https://pytorch.org/docs/stable/generated/torch.full.html) in `full` and `full_like`. To update the standard, I'd want to see upstream support added first.

**Comment by lucascolley** [comment]
Dask also throws a `FutureWarning` since its implementation uses `np.copyto`, which it says may stop working with dask arrays in the future.

**Comment by crusaderky** [comment]
`xp.full` has the same issue.

**Comment by lucascolley** [comment]
@ev-br is this issue blocked by upstream? I know it sounds like we want to add this behaviour to the standard, but until it is added, can't we remove it from array-api-strict?

**Comment by ev-br** [comment]
Yes, exactly. If it's going to be standardized soon, do we want the churn of removing it now and readding it back?
Re adding it to the standard: I think it makes total sense. Without it, one needs something like `full(zeroD_array.item())`, which causes a device synchronization for a GPU backend.

**Comment by lucascolley** [comment]
> If it's going to be standardized soon

Do you know if there is any progress on the PyTorch side?

**Comment by ev-br** [comment]
Sadly, no, I don't.

**Comment by ogrisel** [comment]
So the conclusion it that 0d arrays passed as `fill_value` are only accepted when the pytorch gradient tape is disabled.

So in scikit-learn, we had to "cast" to Python float (`fill_value=float(alpha[0])` instead of `fill_value=alpha[0]`) in one code instance. It's slightly annoying because we need to assume that the dtype is a floating point dtype. For this particular change this is a reasonable assumption. But there might exist other cases where passing an `int`, a `float` or a `bool` might be best, based on the 0d array dtype. Unfortunately, `.item()` is not part of the array API spec:

https://github.com/data-apis/array-api/issues/815

**Comment by ev-br** [comment]
> So this will stay a rare annoyance.

... on par with other annoyances of programming against `-strict`: `array[0]` returning a 0D array not a scalar, the lack of mixed-dtype promotion and the like.
